app/src/main/java/de/westnordost/streetcomplete/data/osm/edits/MapDataWithEditsSource.kt

```diff
@@ -62,8 +63,33 @@ class MapDataWithEditsSource internal constructor(
             val modifiedElements = ArrayList<Pair<Element, ElementGeometry?>>()
             val modifiedDeleted = ArrayList<ElementKey>()
             synchronized(this) {
+                /* We don't want to callOnUpdated if none of the changes affects map data provided
+                 * by MapDataWithEditsSource.
+                 * This is the case if
+                 *  * All keys in deleted are already in deletedElements.
+                 *  * The modified versions of all elements in updated are the same before and after
+                 *    rebuildLocalChanges, except for the timestamp (expected to have few ms
+                 *    difference) and version (never updated locally).
+                 */
+                val deletedUnchanged = deletedElements.containsAll(deleted)
+                val elementsThatMightHaveChanged = updated.mapNotNull { element ->
+                    val key = ElementKey(element.type, element.id)
+                    if (element.isEqualExceptVersionAndTimestamp(updatedElements[key]))
+                        null // we already have the updated version, so this element is unchanged
+                    else
+                        key to get(element.type, element.id) // elementKey and element as provided by MapDataWithEditsSource
+                }
+
```

Comment:
What if `deletedUnchanged` is true and `elementsThatMightHaveChanged` is empty at this point? Why does `rebuildLocalChanges` need to be called then?

Reply:
Does it?
I don't know, and I didn't bother because with cache it's a lot faster than it used to be.
But if it's not necessary maybe we can even return before.

Reply:
okay then let's leave it as it is. Not updating `version` and `timestamp` of the elements may have some unwanted side-effects on upload/conflict solving and quest creation, respectively. Updating the select elements on the other hand is more implementation and testing effort.